src/rules/noAccessStateInSetstateRule.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * @license
+ * Copyright 2018 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Lint from "tslint";
+import { isCallExpression, isClassDeclaration, isPropertyAccessExpression } from "tsutils";
+import * as ts from "typescript";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-access-state-in-setstate",
+        description: "Reports usage of this.state within setState",
+        rationale: Lint.Utils.dedent
+            `Usage of this.state might result in errors when two state calls are \
+            called in batch and thus referencing old state and not the current state.`,
+        options: null,
+        optionsDescription: "",
+        type: "functionality",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING = "Avoid using this.state in first argument of setState.";
```

Comment:
how about we say this instead:
- for uses of `this.state` in the callback update syntax: "References to this.state are not allowed in the setState updater, use the callback arguments instead."
- for uses of `this.state` in the stateChange object syntax: "References to this.state are not allowed in the setState state change object."

Reply:
Good idea! Unfortunately the only way I came up to achieve this was to duplicate the callback function which checks the `setState` argument. Any better idea? :-)

Reply:
I have some ideas for cleaning this up a bit, I can push an additional commit after this merges

Reply:
Maybe you can share your ideas now, since I will refactor this again a bit. Or you just push them, as soon as I provide you guys with the new version!